BLUESPAWN-win-client/src/util/log/JSONSink.cpp

```diff
@@ -0,0 +1,239 @@
+        }
+
+        if(detection->context.note) {
+            detect["note"] = WidestringToString(*detection->context.note);
+        }
+
+        detect["certainty"] = std::to_string(detection->info.GetCertainty());
+        detect["raw-certainty"] = std::to_string(detection->info.GetIntrinsicCertainty());
+
+        if(detection->context.hunts.size()) {
+            detect["associated-hunts"] = json::array();
+            for(const auto& hunt : detection->context.hunts) {
+                detect["associated-hunts"].push_back(WidestringToString(hunt));
+            }
+        }
+
+        detect["associated-data"] = json::object();
+        for(const auto& entry : detection->Serialize()) {
+            detect["associated-data"][WidestringToString(entry.first)] = WidestringToString(entry.second);
+        }
+
+        auto assoc{ detection->info.GetAssociations() };
+        if(assoc.size()) {
+            detect["associated-detections"] = json::array();
+            for(const auto& det : assoc) {
+                detect["associated-detections"].push_back(
+                    { { "strength", static_cast<double>(det.second) }, { "id", std::to_string(det.first->dwID) } });
+            }
+        }
+
+        JSONDoc["bluespawn"]["detections"].push_back(detect);
+    }
+
+    void JSONSink::RecordAssociation(IN CONST std::shared_ptr<Detection>& first,
+                                     IN CONST std::shared_ptr<Detection>& second,
+                                     IN CONST Association& strength) {
+        UpdateCertainty(first);
+        UpdateCertainty(second);
+
+        BeginCriticalSection _{ hGuard };
+
+        if(detections.find(first->dwID) != detections.end()) {
+            AddAssociation(first->dwID, second->dwID, strength);
+        }
+
+        if(detections.find(second->dwID) != detections.end()) {
+            AddAssociation(second->dwID, first->dwID, strength);
+        }
+    }
+
+    void JSONSink::LogMessage(const LogLevel& level, const std::wstring& message) {
+        BeginCriticalSection _{ hGuard };
+
+        if(level.Enabled()) {
+            std::map<std::string, std::string> msg = {};
+            if(level.detail) {
+                msg["detail"] = *level.detail == Detail::High     ? "high" :
+                                *level.detail == Detail::Moderate ? "moderate" :
+                                                                    "low";
+            }
+            SYSTEMTIME time{};
+            GetLocalTime(&time);
+            msg["time"] = WidestringToString(FormatWindowsTime(time)).c_str();
+            msg["message"] = WidestringToString(message).c_str();
+            JSONDoc["bluespawn"]["log-messages"].push_back(msg);
+        }
+    }
+
+    bool JSONSink::operator==(const LogSink& sink) const {
+        return (bool) dynamic_cast<const JSONSink*>(&sink) &&
+               dynamic_cast<const JSONSink*>(&sink)->wFileName == wFileName;
+    }
+
+    void JSONSink::Flush() {
```

Comment:
Needs to be regularly called during monitor mode

Reply:
my code here looks the same as what's in XMLSink?

Reply:
Oh, I might have missed it last time, but you needed the thread watching for the flushlogs event and calling this. You have it now, at least.